Clarify README section one based on build w/o credentials (#1384, #1370) (#1385)
DataBiosphere/azul: README.md

via a plug-in mechanism, allowing the Boardwalk UI codebase to be functionally
generic with minimal need for project-specific behavior.

## 1.2 Architecture Diagram

![Azul architecture diagram](docs/azul-arch.png)

# 2. Getting Started

## 2.1 Development Prerequisites

- Python 3.6 (3.7 does not work) with `pip`. Micro versions 3.6.5 and 3.6.8 are 

>   known to work, but 3.6.9 does not.

Reviewer: I'm using 3.6.9. What exactly does not work?

Author: Failure to install dependencies via pip due to SSL errors.

Author: The issue may be specific to Mac/OSX.

Reviewer: Who has observed the problem under 3.6.9 on OS X? If it is specific to OS X, this fact should be stated here.  I would like this issue with 3.6.9 to be corroborated/demonstrated. If that's not possible, I would like to remove the clause from this sentence. If it is possible, I'd like a ticket to be created and a reference to the ticket put into the README right here. 3.6 is our target platform so it's a big deal if Azul does not work on the most recent 3.6 release.

Author: I encountered it on OSX (on the macbook I was loaned when I first started here). Downgrading to 3.6.5 fixed the issue on that computer. I haven't tried using 3.6.9 on any other platform, so I can't verify that it's specific.

Reviewer: Did you install 3.6.9 on that loaner or was it already on there?

Author: I'm fairly certain I installed it. Only 2.7 was installed by default.

- The `bash` shell

- Docker for running the tests. The minimal required version is uncertain but

>   19.03, 18.09, and 17.09 are known to work.

Reviewer: As we discovered today, you seem to have issues with `make format`. Let's see if those have anything to do with your use of 19.03.  Then again I'm using 19.03 as well and `make format` seems to work.

Author: I downgraded to 18.09 and the error persists.

   Newer versions of pip are incompatible with some of our requirements, hence
   the pin on the 10.0.1 version.
   

>    If requirements fail to build, try installing `python3-dev`.

Reviewer 2: Please add that this only applies to Ubuntu. I think putting the whole
```
apt install python3-dev
```
command would also be helpful.

   Newer versions of pip are incompatible with some of our requirements, hence
   the pin on the 10.0.1 version.
   

>    If requirements fail to build, Ubuntu users should try installing 

Reviewer: ```suggestion
   Ubuntu users of the standard Python that Ubuntu ships with must install 
```

6. Run `make`. It should say `Looking good!` If one of the sanity checks fails,
   address the complaint and repeat. The various sanity checks are defined in
   `common.mk`.
   
   Some checks utilize git. If the repository was cloned using http, you may
   be asked to authenticate. Using ssh is recommended, but for the purposes of
   this check the credentials may be left blank.

> 7. Make sure Docker is running, and not as root. Ubuntu users may need to add 

Reviewer 2: I think just this first sentence is sufficient. It's possible the procedures for setting up Docker on Ubuntu will change, and it's easier to maintain if we assume the user will be able to figure out how to do this themselves. Do you agree?

I think pointing to Docker's install documentation is also a reasonable alternative here.

6. Run `make`. It should say `Looking good!` If one of the sanity checks fails,
   address the complaint and repeat. The various sanity checks are defined in
   `common.mk`.
   
   Some checks utilize git. If the repository was cloned using http, you may
   be asked to authenticate. Using ssh is recommended, but for the purposes of
   this check the credentials may be left blank.

> 7. Make sure [Docker is installed](https://docs.docker.com/v17.12/install/) 

Reviewer: This should be in the `Development Prerequisites` section.

Reviewer: And don't link to a specific version. Or link to all supported versions.

6. Run `make`. It should say `Looking good!` If one of the sanity checks fails,
   address the complaint and repeat. The various sanity checks are defined in

>    `common.mk`. The github credentials may be left blank.

Reviewer 2: Since the prompt for Github credentials is dependent on how you cloned the repository and authenticate with Github, I would recommend changing this to something more general.

Perhaps, "Some of the checks make use of git, and if you cloned the repository with via http, you may be asked to authenticate. We recommend using ssh, but for the purposes of this check you can leave the credentials blank."

Feel free to change these words to your taste.

6. Run `make`. It should say `Looking good!` If one of the sanity checks fails,
   address the complaint and repeat. The various sanity checks are defined in
   `common.mk`.
   

>    Some checks utilize git. If the repository was cloned using http, you may

Reviewer: ```suggestion
   Some checks utilize Git. If the repository was cloned using http, you may
```

or 

```suggestion
   Some checks utilize `git`. If the repository was cloned using HTTP, you may
```

Acronyms all-caps.

>    be asked to authenticate. Using ssh is recommended, but for the purposes of

Reviewer: ```suggestion
   be asked to authenticate. Using SSH is recommended, but for the purposes of
```

   this check the credentials may be left blank.

> 7. Make sure Docker is running without root access. 

Reviewer: Line wraps prematurely it seems.

python scripts/envhook.py install
```

> If `envhook.py` is already present, it must be renamed or removed before the

Reviewer 2: ```suggestion
The script works by adding a `sitecustomize.py` file to your virtual environment. If a different `sitecustomize` module is already present in your Python path, it must be renamed or removed before the
```

python scripts/envhook.py install
```

The script works by adding a `sitecustomize.py` file to your virtual 
environment. If a different `sitecustomize` module is already present in your 

> Python path, its `sitecustomize.py` file must be renamed or removed before the 

Reviewer: @jessebrennan I thought you implemented a check for that?

Reviewer 2: There is a check. The command warns you if there is already a sitecustomize. These instructions are still helpful though, if you lack the context of what sitecustomize is, or are unsure how to find the file.

Reviewer: OK

python scripts/envhook.py install
```

The script works by adding a `sitecustomize.py` file to your virtual 
environment. If a different `sitecustomize` module is already present in your 
Python path, its `sitecustomize.py` file must be renamed or removed before the 
installation can proceed. The current install location can be found by importing
`sitecustomize` and inspecting the module's `__file__` attribute.

Whether you installed `envook.py` or not, a couple more steps are necessary to
configure PyCharm for Azul:

1. Under *Settings* -> *Project—Interpreter* select the virtual environment
   created above.

2. Set the `src` and `test` folders as source roots by right-clicking each 
   folder name and selecting *Mark Directory as* → *Sources Root*.
   
3. Exclude the `.venv`, `lambdas/indexer/vendor`, and  `lambdas/service/vendor` 
   folders by right-clicking each folder name and selecting *Mark Directory as*

>    → *Excluded*.

Author: #1370

Also see https://chalice.readthedocs.io/en/latest/topics/packaging.html

# 12. Development tools

> ## 12.1 OpenAPI Development

Reviewer: ```suggestion
## 12.1 OpenAPI development
```
